tests/run/pep3135_class_cell.py

```diff
@@ -0,0 +1,116 @@
+# cython: language_level=3
+# mode: run
+# tag: pep3135, pure3.0
+
+
+class C(object):
+    """
+    >>> obj = C()
+    >>> obj.method_1()
+    2
+    >>> obj.method_2()
+    3
+    >>> obj.method_3()
+    ['__class__', 'self']
+    >>> obj.method_4()
+    ['self']
+    """
+
+    @classmethod
+    def class_method(cls):
+        return 2
+
+    @staticmethod
+    def static_method():
+        return 3
+
+    def method_1(self):
+        return __class__.class_method()
+
+    def method_2(self):
+        return __class__.static_method()
+
+    def method_3(self):
+        __class__
+        return sorted(list(locals().keys()))
+
+    def method_4(self):
+        return sorted(list(locals().keys()))
+
+
+class D:
+    """
+    >>> obj = D()
+    >>> obj.method(1)
+    1
+    >>> obj.method(0)
+    Traceback (most recent call last):
+    ...
+    UnboundLocalError: local variable '__class__' referenced before assignment
+    """
+    def method(self, x):
+        if x: __class__ = x
+        print(__class__)
+
+
+class E:
+    """
+    >>> obj = E()
+    >>> obj.method()().__name__
+    'E'
+    """
+    def method(self):
+        def inner(): return __class__
+        return inner
+
+
+class F:
+    """
+    >>> obj = F()
+    >>> obj.method()()().__name__
+    'F'
+    """
+    def method(self):
+        def inner():
+            def inner_inner():
+                return __class__
+            return inner_inner
+        return inner
+
+
+class G:
+    """
+    >>> obj = G()
+    >>> obj.method().__name__
+    'H'
+    """
+    def method(self):
+        class H:
+            def inner(self):
+                return __class__
+        return H().inner()
+
+
+class I:
+    """
+    >>> obj = I()
+    >>> obj.method()()().__name__
+    'J'
+    """
+    def method(self):
+        def inner():
+            class J:
+                def inner(self):
+                    return __class__
+            return J().inner
+        return inner
+
+
+class K:
+    """
+    >>> OldK = K
+    >>> K = None
```

Comment:
I don't think that this really changes the global name *in the module*, where the method looks it up. Please move these assignments into the main module code, just after defining the class `K`.

Reply:
Ok, what I meant was that this should work:
```
class K(object):
    """
    >>> OldK().method().__name__
    'K'
    """
    def method(self): return __class__

OldK = K
K = None
```
It probably doesn't currently.

Reply:
Yeah, that currently breaks with the predictable `'NoneType' object has no attribute '__name__'`.

Reply:
So what's happening here is that `K` is being rebound to `None` in the outer scope, but the `DefNode`'s scope should hold onto the reference, if I have that right.